features/FxA-56-email-deliverability/README.md

```diff
@@ -1,66 +1,160 @@
+This will help us to determine if any particular type of email
+is more likely to be rejected, which may indicate some sort of
+email-content-related issue.
 
-If the email is not valid and keeps bouncing then we remove it from our system or at least stop sending emails to it.
+### Bounce and complaint rate by domain
 
-* Check domain MX Records
+We will produce a graph showing the number of bounce notifications per day,
+as a percentage of the number of emails sent that day,
+broken down by email domain.
+It will look something like this:
+
+![Mock graph for bounce mail by domain](bounce_rate_by_domain.jpg)
+
+We will produce a similar graph for complaint notifications.
+
+This will help us to determine if we're having deliverability issues
+with a specific provider, which will aid in debugging.
+
+
+
+----
+
+## Detailed design
 
-We need to add a way to check against bad MX records when users sign up for new accounts in FxA.
+We will implement the following different parts 
+to help us debug our email deliverability problems:
+
+* [ ] Add flow events for email bounce notifications from SES
+* [ ] Add flow events for email complaint notifications from SES
+* [ ] Add flow events for email delivery notifications from SES
+* [ ] Add top-N domain bucketing to email bounce and complaint event logs
+* [ ] Create the "completion rate for email flows" dashboard
+* [ ] Create the "lost mail rate for email flows" dashboard
+* [ ] Create the "bounce rate by template" dashboard
+* [ ] Create the "complaint rate by template" dashboard
+* [ ] Create the "bounce rate by domain" dashboard
+* [ ] Create the "complaint rate by domain" dashboard
+
+To support reporting of bounce rate by domain
+while respecting user privacy,
+we will identify a short list of the most popular domains.
+Events for email addresses outside of this list
+will be grouped together as an "other" category.
+
+With these metrics in place we will try
+one simple tactic for improving deliverability -
+rejecting further emails to addresses that have recently bounced.
+This should improve our sender reputation
+and also give the user more opportunity
+to correct an incorrectly-entered email address.
+
+We will implement this by:
+
+* [ ] Maintain a database table of all bounce and complain notifications,
+      keyed by email address.
+* [ ] Expire entries from the bounced-emails table after 1 week.
+* [ ] When a user requests email be sent to an address that had a soft
+      bounce notification in the last 5 minutes, return a new "soft failure"
+      error code and show appropriate messaging.
+* [ ] When a user requests email be sent to an address that had a hard
+      bounce notification in the last 24 hours, return a new "hard failure"
+      error code and show appropriate messaging.
+* [ ] When a user requests email be sent to an address that had a complaint
+      notification in the last 24 hours, return a new "complaint received"
+      error code and show appropriate messaging.
```

Comment:
Just to be clear on definitions:
- soft bounce = auto-responder (e.g. vacation, thanks for contacting support, daily limit reached, etc.)
- hard bounce = mx record doesn't exist, user at email provider doesn't exist or inbox full.

Anything I'm missing? I feel like my soft bounce is missing something. If I am not mistaken, there is not much risk at sending again on soft bounces. It becomes more a question of ignoring them after a while since they provide no value if they continuously soft bounce.

On the other hand, I permanently banned hard bounces because blocking them forever did less harm then bouncing a second time on the same address (which is actually pretty bad when done frequently).

Perhaps as a phase 1, we permanently block the hard bounces but we figure out for a later phase a way to unblock them. (e.g. send us an email from this account)